ddl/cache.lua

```diff
@@ -0,0 +1,136 @@
+local fiber = require('fiber')
+
+local cache = nil
+
+local CACHE_LOCK_TIMEOUT  = 3
+local SPACE_NAME_IDX      = 1
+local SHARD_FUNC_NAME_IDX = 2
+local SHARD_FUNC_BODY_IDX = 3
+
+-- Function decorator that is used to prevent cache_build() from being
+-- called concurrently by different fibers.
+local function locked(f)
+    return function(...)
+        local ok = cache.lock:put(true, CACHE_LOCK_TIMEOUT)
+
+        if not ok then
+            error("cache lock timeout is exceeded")
+        end
+
+        local status, err = pcall(f, ...)
+        cache.lock:get()
+
+        if not status or err ~= nil then
+            return err
+        end
+    end
+end
+
+-- Build cache.
+--
+-- Cache structure format:
+-- cache = {
+--   spaces = {
+--     'space_name' = {
+--       raw = {},       -- raw sharding metadata, used for ddl.get()
+--       processed = {}, -- table with parsed dot notation (like {'foo', 'bar'})
+--                       -- or a function ready to call (or a string with an error)
```

Comment:
A kind of nitpicking or thoughts on the code. How about the following structure?

```lua
{
    error = <string> or <nil>,
    raw_tuple = <tuple object> (<nil> at error),
    callable = <function> or <nil>,
    parsed_func_name = <table> or <nil>,
}
```

It avoid polymorph fields and may be simpler to understand and use.

I don't mind current structure: it solves given task. If you disagree, feel free to close this comment.

Reply:
reworked, done